### PR TITLE
Loyalty program give reward points

### PR DESCRIPTION
**Pull request by dvinubius**
Created for discussion

### CODE REVIEW

File: contracts/LoyaltyCardMaster.sol [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+    }
+
+     /**
+      @notice Add loyalty points to a given IF user account
+      @param account The IF user account
+      @param points Number of points to add
+     */
+    function addPointsAccount(address account, uint256 points)
+        external 
+        onlyOperator
```
Comment (friday-james) [inline contracts/LoyaltyCardMaster.sol:217]: May remove the `onlyOperator` modifier from this function

File: test/test-LoyaltyCardMaster.ts [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+  })
+
+  it('Should recognize transferred tokens as staked', async function () {
+    // "staked" means transferred to a whitelisted destination
+    await loyaltyCardMaster.setMinter(owner.address)
+    await loyaltyCardMaster.mint(user.address)
+    const mintedTokenId = await loyaltyCardMaster.originalOwnerToTokenId(
+      user.address
+    )
+    await loyaltyCardMaster.addOperator(operator1.address)
```
Comment (friday-james) [inline test/test-LoyaltyCardMaster.ts:375]: `addOperator` is not needed

File: contracts/LoyaltyCardMaster.sol (hunk 1 above)
Comment (acceleratedmonkey) [reply contracts/LoyaltyCardMaster.sol:217]: why tho? i thought only operator can add points?

Comment (friday-james) [reply contracts/LoyaltyCardMaster.sol:217]: Sorry, I forgot to add an explanation 😆  
The function `_addPointsAccount` already has the `onlyOperator` modifier. Both `addPointsCard` and `addPointsAccount` are calling `_addPointsAccount`, but `addPointCard` doesn't have the modifer. I think `addPointsAccount` doesn't need it as well.

File: test/test-LoyaltyCardMaster.ts (hunk 2 above)
Comment (dvinubius) [reply test/test-LoyaltyCardMaster.ts:375]: 👍

**Comment by acceleratedmonkey** [comment]
pls exclude the MasterContractV2 from the PR

**Comment by dvinubius** [comment]
@acceleratedmonkey 👍 just did, sorry for that

File: contracts/LoyaltyCardMaster.sol [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+      @dev Override the ERC721 default in order to block transfers to non-whitelisted destinations
+     */
+    function _transfer(
+        address from,
+        address to,
+        uint256 tokenId
+    ) internal override {
+        // this only goes through if token is being sent to a WL destination
+        // or a WL destination (related to Impossible Finance) transfers it (back)
+        if (!whitelistedDestination[to] && !whitelistedDestination[from])
```
Comment (acceleratedmonkey) [inline contracts/LoyaltyCardMaster.sol:290]: or, should we instead use || opeartion here, we have the similar scenario in vidia where we whitelist the smart contract for staking purposes, 
https://github.com/ImpossibleFinance/launchpad-contracts/blob/8a8e8ea2bcce4c6a85586e2eba47a69b60255d01/contracts/vIDIA.sol#L386

the reason for using the || operator is that in the case of unstaking from the contract if you have to whitelist both side you will need to whitelist every single owner of card

**Comment by friday-james** [comment]
I think this is good to go. @acceleratedmonkey sir any thoughts?

**Review by acceleratedmonkey — APPROVED** [review]

**Pull request merged**